### PR TITLE
Tile and distribute linalg.generic in DispatchLinalgOnTensor

### PR DESCRIPTION
**Pull request by ThomasRaoux**

### CODE REVIEW

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:330]: Misleading comment. the `ConvertToGPU` pass does not assume a tile size right? I am confused cause the launch config isnt even called in the `ConvertToGPU` Pass.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:336]: Could you add comments to whats happening here. Hard to parse this a bit.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:615]: Oh Nice! This was exactly what I was trying to do. Should work well with Hanhan's PR #5136

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:855]: So the only reason we have the `MakeDispatchWorkgroupsOp` pattern is for the fall back right? In other words in the `ConvertToGPU` pass, the mapping of iterations of `linalg` operation to global invocation ID is not used in the Linalg to tensors path. That would be great, cause then we can move all that loop distribution logic all the way in Flow.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Looks awesome! I am super excited about this landing! It address a lot of hidden issues by enabling this path.

Just a nit on comments for setting the launch config. Once it is addressed, it is good to go.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (ThomasRaoux) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp]: Good catch, you're right I had the comment backward. The problem is that when vectorization is off the second level of tiling is skipped and we end up mapping one element to one thread. So the real problem is that if we picked something different in LaunchConfig the number of workgroup is wrong. I updated the comment, hopefully this is clear.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp
Comment (ThomasRaoux) [reply iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:855]: Correct, there is still few cases that don't tile, I can look at those at some point. Yes for all those cases the `ConverToGPU` global invocation ID is not reached. It would be good to do the loop distribution in flow indeed. 
The main case I have seen remaining is when the linalg affine map are not projected permutation.

**Comment by ThomasRaoux** [comment]
Thanks Mahesh, please take another look.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:340]: Please add a TODO here. FOr the linalg on tensors path this could actually be anything other than 1. It can be even on the old path with a few changes, but why bother when its going away. For now, leave a TODO and come back to it when we deprecate the old path.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:759]: It might just be getting mixed with the old paths. This should not be needed actually.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:758]: When we drop the old path, this will have to go through the new path.

**Review by MaheshRavishankar — APPROVED** [review]

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (ThomasRaoux) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:759]: I still need this check otherwise some linalg op with weird affine map crash during tiling. I can look at why this happen after if you think this is not expected.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (ThomasRaoux) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:758]: we still need this even in the new path right?

**Comment by ThomasRaoux** [comment]
I'll merge this PR, if there are still concerns with your comments I can address it after.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:759]: Sorry, bad comment. I was saying that this is basically saying dont do anything with 0-rank operations. But those need to be handled as well (i.e. lowered to loops albeit no-op) at some point. I am guessing this isnt tested on the new path. 
One thing to do here is just test what happens if a `linalg.generic` is just  0-rank operations.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:760]: Yeah, this is not expected. I dont see why it would carsh (at least would be good to know the affine map. Is this because of `reverse`? If so worth triaging upstream.

**Comment by hanhanW** [comment]
This seems to break compiling mobilenet_v2 in iree-android-benchmark.

To repro:
```
iree-translate --iree-mlir-to-vm-bytecode-module --iree-hal-target-backends=vulkan-spirv ~/mobilenet_v2.mlir -o a.vmfb --iree-spirv-enable-vectorization --iree-vulkan-target-triple=valhall-g77-unknown-android10
```

The mobilenet_v2.mlir can be downloaded from https://storage.googleapis.com/iree-model-artifacts/mobilenet-v2.tar.gz

It fails in ConvertToSPIRVPass. The input IR of ConvertToSPIRVPass: https://gist.github.com/hanhanW/1f1c5cb70ff1486c04783ea052e2703d

The error message is `failed to legalize operation 'vector.broadcast'`

**Comment by antiagainst** [comment]
Oh, a bunch of size-1 vectors are generated. Those aren't supported by SPIR-V. (They should just be scalars.)

@hanhanW: This is still checking the existing path? I think we should switch the benchmark tests to go via the new path. 

Also it would be nice to check in a faked weights MobileNet v2. Discovering breakage via benchmarking tests is not that good. :) 
 
@ThomasRaoux: can we revert this for now and have the issues fixed before landing it?

**Comment by ThomasRaoux** [comment]
> Oh, a bunch of size-1 vectors are generated. Those aren't supported by SPIR-V. (They should just be scalars.)
> 
> @hanhanW: This is still checking the existing path? I think we should switch the benchmark tests to go via the new path.
> 
> Also it would be nice to check in a faked weights MobileNet v2. Discovering breakage via benchmarking tests is not that good. :)
> 
> @ThomasRaoux: can we revert this for now and have the issues fixed before landing it?

Yes, let me revert it and I'll fix it offline.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (ThomasRaoux) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:760]: Mahesh, yes the problem is with `reverse`, I'll work on fixing this part in a second step.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (ThomasRaoux) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:759]: Good point, I can add a test for that.

**Pull request merged**
